src/github.com/matrix-org/dendrite/clientapi/sync/requestpool.go

```diff
@@ -54,34 +75,70 @@ func (rp *RequestPool) OnIncomingSyncRequest(req *http.Request) util.JSONRespons
 		"timeout": timeout,
 	}).Info("Incoming /sync request")
 
-	res, err := rp.currentSyncForUser(syncReq)
-	if err != nil {
-		return httputil.LogThenError(req, err)
-	}
-	return util.JSONResponse{
-		Code: 200,
-		JSON: res,
+	// Fork off 2 goroutines: one to do the work, and one to serve as a timeout.
+	// Whichever returns first is the one we will serve back to the client.
+	// TODO: Currently this means that cpu work is timed, which may not be what we want long term.
+	timeoutChan := make(chan struct{})
+	timer := time.AfterFunc(timeout, func() {
+		close(timeoutChan) // signal that the timeout has expired
+	})
+
+	done := make(chan util.JSONResponse)
+	go func() {
+		syncData, err := rp.currentSyncForUser(syncReq)
```

Comment:
There's currently no cancellation mechanism at all built into the actual `/sync` work. This was done on purpose, because whilst we could reasonably recheck the exit condition on `rp.cond.Wait()` what we actually want is to be able to shoot the entire request in the head if we've already timed out. That's the entire point of using `Context` objects, where we can periodically check `ctx.Err()` but to add this into this PR was getting excessive, so I punted it.